src/move/emojicoin_dot_fun/sources/tests.move

```diff
@@ -2218,114 +2297,601 @@
+        abort_code = emojicoin_dot_fun::emojicoin_dot_fun::E_NO_MARKET,
+        location = emojicoin_dot_fun::emojicoin_dot_fun,
+    )] fun provide_liquidity_no_market() {
+        provide_liquidity<BlackCatEmojicoin, BlackCatEmojicoinLP>(
+            &get_signer(USER),
+            @black_cat_market,
+            1,
+        );
+    }
+
+    #[test, expected_failure(
+        abort_code = emojicoin_dot_fun::emojicoin_dot_fun::E_LIQUIDITY_NO_QUOTE,
+        location = emojicoin_dot_fun::emojicoin_dot_fun,
+    )] fun provide_liquidity_no_quote() {
+        init_package_then_exact_transition();
+        provide_liquidity<BlackCatEmojicoin, BlackCatEmojicoinLP>(
+            &get_signer(USER),
+            @black_cat_market,
+            0,
+        );
+    }
+
+    #[test, expected_failure(
+        abort_code = emojicoin_dot_fun::emojicoin_dot_fun::E_STILL_IN_BONDING_CURVE,
+        location = emojicoin_dot_fun::emojicoin_dot_fun,
+    )] fun provide_liquidity_still_in_bonding_curve() {
+        init_package_then_simple_buy();
+        provide_liquidity<BlackCatEmojicoin, BlackCatEmojicoinLP>(
+            &get_signer(USER),
+            @black_cat_market,
+            1
+        );
+    }
+
+    #[test] fun provide_liquidity_with_truncation() {
+        // Invoke an exact transition, then have simple buy user buy against the CPAMM.
+        init_package_then_exact_transition();
+        timestamp::update_global_time_for_test(EXACT_TRANSITION_TIME + 1);
+        mint_aptos_coin_to(SIMPLE_BUY_USER, SIMPLE_BUY_INPUT_AMOUNT);
+        swap<BlackCatEmojicoin, BlackCatEmojicoinLP>(
+            &get_signer(SIMPLE_BUY_USER),
+            @black_cat_market,
+            SIMPLE_BUY_INPUT_AMOUNT,
+            SWAP_BUY,
+            SIMPLE_BUY_INTEGRATOR,
+            SIMPLE_BUY_INTEGRATOR_FEE_RATE_BPS,
+        );
+
+        // Get the new CPAMM reserves.
+        let  ( _, _, _, cpamm_real_reserves, _, _, _, _, _, _, _, _, _) = unpack_market_view(
+            market_view<BlackCatEmojicoin, BlackCatEmojicoinLP>(@black_cat_market)
+        );
+        let (base, quote) = unpack_reserves(cpamm_real_reserves);
+        let cpamm_real_reserves = MockReserves { base, quote };
+
+        // Declare new amount of quote to provide.
+        let quote_amount = SIMPLE_BUY_INPUT_AMOUNT / 10;
+
+        // Get base amount, for truncation present during proportion calculation.
+        let proportion_numerator =
+            (cpamm_real_reserves.base as u128) * (quote_amount as u128);
+        let proportion_denominator = (cpamm_real_reserves.quote as u128);
+        assert!(proportion_numerator % proportion_denominator != 0, 0);
+        let base_amount = ((proportion_numerator / proportion_denominator) as u64) + 1;
+
+        // Determine amount of liquidity tokens provided.
+        let lp_coin_amount = ((
+            (quote_amount as u128) * (get_LP_TOKENS_INITIAL() as u128) /
+            (cpamm_real_reserves.quote as u128)
+        ) as u64);
```

Comment:
This calculation corresponds to one of the formulas in the blackpaper right? If it is, it'd be nice to note which formula it is in a comment to make it more clear why this should be the expected value

Reply:
Addressed in b4f5b4d